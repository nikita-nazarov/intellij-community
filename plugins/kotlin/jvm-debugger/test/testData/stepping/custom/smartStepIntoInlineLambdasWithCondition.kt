// Copyright 2000-2023 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package smartStepIntoInlineLambdasWithCondition

fun foo1() = 1
fun foo2() = 2
fun foo3() = 3

inline fun foo(a: Boolean, b: Boolean, c: Boolean) {
    1.letIf(a) { foo1() }.letIf(b) { foo2() }.letIf(c) { foo3() }
}

inline fun <T, R> T.letIf(condition: Boolean, block: (T) -> R?): R? {
    return if (condition) block(this) else null
}

fun main() {
    // STEP_INTO: 1
    // SMART_STEP_INTO_BY_INDEX: 2
    // STEP_OUT: 2
    // SMART_STEP_INTO_BY_INDEX: 4
    // STEP_OUT: 2
    // SMART_STEP_INTO_BY_INDEX: 6
    // RESUME: 1
    //Breakpoint!
    foo(true, true, false)

    // STEP_INTO: 1
    // SMART_STEP_INTO_BY_INDEX: 2
    // STEP_OUT: 2
    // SMART_STEP_INTO_BY_INDEX: 4
    // STEP_OUT: 2
    // SMART_STEP_INTO_BY_INDEX: 6
    // RESUME: 1
    //Breakpoint!
    foo(false, false, false)

    // STEP_INTO: 1
    // SMART_STEP_INTO_BY_INDEX: 2
    // STEP_OUT: 2
    // SMART_STEP_INTO_BY_INDEX: 4
    // STEP_OUT: 2
    // SMART_STEP_INTO_BY_INDEX: 6
    // RESUME: 1
    //Breakpoint!
    foo(false, true, false)
    foo(false, true, true)
    foo(true, false, true)
    foo(true, true, true)
}
