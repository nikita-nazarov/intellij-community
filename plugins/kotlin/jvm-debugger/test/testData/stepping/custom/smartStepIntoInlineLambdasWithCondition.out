LineBreakpoint created at smartStepIntoInlineLambdasWithCondition.kt:21
LineBreakpoint created at smartStepIntoInlineLambdasWithCondition.kt:31
LineBreakpoint created at smartStepIntoInlineLambdasWithCondition.kt:41
Run Java
Connected to the target VM
smartStepIntoInlineLambdasWithCondition.kt:21
smartStepIntoInlineLambdasWithCondition.kt:5
smartStepIntoInlineLambdasWithCondition.kt:5
smartStepIntoInlineLambdasWithCondition.kt:9
smartStepIntoInlineLambdasWithCondition.kt:31
smartStepIntoInlineLambdasWithCondition.kt:41
Disconnected from the target VM

Process finished with exit code 0
